spec/datadog/core/telemetry/metric_queue_spec.rb

```diff
@@ -0,0 +1,128 @@
+require 'spec_helper'
+
+require 'datadog/core/telemetry/metric_queue'
+require 'datadog/core/telemetry/metric'
+
+RSpec.describe Datadog::Core::Telemetry::MetricQueue do
+  let(:metric_queue) { described_class.new }
+  let(:metric_klass) { Datadog::Core::Telemetry::Metric::Count }
+  subject(:empty_metrics_queue) do
+    {
+      'generate-metrics' => {},
+      'distributions' => {},
+    }
+  end
+
+  describe '#add_metric' do
+    context 'no previous metric' do
+      it 'creates a new metric entry point and stores it' do
+        expect(metric_queue.send(:metrics)).to eq(empty_metrics_queue)
+        expect(metric_klass).to receive(:new).with('test_metric_name', { foo: :bar }).and_call_original
+        expect_any_instance_of(metric_klass).to receive(:update_value).with(1).and_call_original
+        metric_queue.add_metric(
+          'test_namespace',
+          'test_metric_name',
+          1,
+          { foo: :bar },
+          metric_klass
+        )
+
+        expect(metric_queue.send(:metrics)[metric_klass.request_type]['test_namespace']).to_not be_nil
+        expect(metric_queue.send(:metrics)[metric_klass.request_type]['test_namespace']['test_metric_name']).to_not be_nil
+
+        metric_instace = metric_queue.send(:metrics)[metric_klass.request_type]['test_namespace']['test_metric_name']
+        expect(metric_instace).to be_a(metric_klass)
```

Comment:
Minor: Since the last check won't pass unless the previous ones do, I suggest keeping only the last check. Especially since in this case we're asserting on private state of the class, it's IMHO a good idea to minimize the number of assertions, so it's easier to fix if we need to tweak the representation.